### PR TITLE
OCPBUGS-19817, SDN-4162: Upgrade IPsec with single daemonset pod

### PR DESCRIPTION
**Pull request by pperiyasamy**
This PR ensures IPsec daemonset pod uses IPsec packages installed on the host and it can be used to configure IPsec for both east-west and north-south traffic. It also seamlessly handles both fresh install and upgrade scenario so that network operator bring up only one IPsec pod on every node.

### CODE REVIEW

File: pkg/platform/platform.go
Comment (jcaamano) [inline pkg/platform/platform.go:153]: Why do you need this? You just added the Machine configs in this PR. Please, explain and split on its own commit because I don't see how this is related to the commit you added it.

**Review by jcaamano — COMMENTED** [review]
@pperiyasamy You changed strategy multiple times throughout the commits. At this point I think it is best to squash. Except the last commit, partially, where you check if the ipsec extensions exist under different name. That merits is own commit with its own explanation, because the machine configs were introduced in this PR so it is a strange thing to do.

Also please, add a big comment explaining the overall strategy of the different enablement paths. Otherwise it is difficult to follow.

File: pkg/network/ovn_kubernetes.go
Comment (pperiyasamy) [reply pkg/network/ovn_kubernetes.go:336]: This again needs if checks for c`.IPsecConfig != nil `and `bootstrapResult.OVN.IPsecUpdateStatus.IsIPsecUpgrade`, hence not making changes into this logic.

File: pkg/bootstrap/types.go
Comment (pperiyasamy) [reply pkg/bootstrap/types.go:107]: no, i've not talked to @pliurh about this, would talk to him by this week.

File: pkg/platform/platform.go
Comment (pperiyasamy) [reply pkg/platform/platform.go:164]: added with commit 239d296741c5f329b05d83f055c964f24d16ec27 and doc comment in the code.

File: pkg/network/ovn_kubernetes.go
Comment (pperiyasamy) [reply pkg/network/ovn_kubernetes.go:1672]: as discussed, we still need  `ovn-ipsec` daemonset yaml file to render with `CreateWaitAnnotation`  until it cleans up previous IPsec config and then upgrade to latest IPsec config.

File: pkg/network/ovn_kubernetes.go
Comment (pperiyasamy) [reply pkg/network/ovn_kubernetes.go:1671]: For 4.15, ovn-ipsec-host is used only when both EW+NS set.

File: pkg/network/ovn_kubernetes.go
Comment (pperiyasamy) [reply pkg/network/ovn_kubernetes.go]: good point! done.

File: pkg/network/ovn_kubernetes.go
Comment (pperiyasamy) [reply pkg/network/ovn_kubernetes.go]: sure, done.

File: pkg/network/ovn_kubernetes.go
Comment (pperiyasamy) [reply pkg/network/ovn_kubernetes.go]: yes, done

File: pkg/network/ovn_kubernetes.go
Comment (pperiyasamy) [reply pkg/network/ovn_kubernetes.go]: correct. removed all debug logs.

File: pkg/network/ovn_kubernetes.go
Comment (pperiyasamy) [reply pkg/network/ovn_kubernetes.go]: added comments in the code, hope it's bit clear now.

File: pkg/network/ovn_kubernetes.go
Comment (pperiyasamy) [reply pkg/network/ovn_kubernetes.go:588]: as discussed, added two commits for 4.14->4.15 and pre-4.14->4.15 upgrade paths. 
i hope things are clear with this.

File: pkg/apply/apply.go
Comment (pperiyasamy) [reply pkg/apply/apply.go]: yes, i think we can just use `names.CreateWaitAnnotation`.  made relevant changes.

File: pkg/network/ovn_kubernetes.go
Comment (pperiyasamy) [reply pkg/network/ovn_kubernetes.go]: sure, done.

File: pkg/network/ovn_kubernetes.go
Comment (pperiyasamy) [reply pkg/network/ovn_kubernetes.go]: There is no need of IPsec daemonset in running state for enabling IPsec in OVN.
But it must be in running state to process disable IPsec option in OVN which is being handled in the PR.

File: pkg/network/ovn_kubernetes.go
Comment (pperiyasamy) [reply pkg/network/ovn_kubernetes.go]: removed rendering for IPsecMachineConfig from the PR, would reintroduce when new `ipsecExternal` API support is included.

File: pkg/network/ovn_kubernetes.go
Comment (pperiyasamy) [reply pkg/network/ovn_kubernetes.go]: we disable ipsec only for EW temporarily, no issues noticed from upgrade testing (I think traffic must flow with ipsec at that time).

File: pkg/network/ovn_kubernetes.go
Comment (pperiyasamy) [reply pkg/network/ovn_kubernetes.go]: yes, done.

File: pkg/network/ovn_kubernetes.go
Comment (pperiyasamy) [reply pkg/network/ovn_kubernetes.go]: sure, moved out to a new function.

**Comment by kyrtapz** [comment]
This PR doesn't work with hypershift:
```
I1128 11:31:50.306160       1 log.go:194] Failed to validate Network.Spec: failed to get ipsec machine config for master: no matches for kind "MachineConfig" in version "machineconfiguration.openshift.io/v1"
```
AFAIK HyperShift does not use machineconfiguration.openshift.io/v1 APIs, so the approach taken here wouldn't work there.

**Comment by pperiyasamy** [comment]
> AFAIK HyperShift does not use machineconfiguration.openshift.io/v1 APIs, so the approach taken here wouldn't work .

@kyrtapz Would [this](https://github.com/openshift/cluster-network-operator/pull/2087/commits/041e7e32fbbd5e5a886a95a1b5a4e4616ae061a8#diff-13e60e30d72f7b5c263fabc48ae9533c32531df3983d7e70d8a4e1a0a1c1a97dR138-R167) unblock HyperShift considering IPsec for NS is never supported there ?

File: pkg/network/ovn_kubernetes.go
Comment (jcaamano) [inline pkg/network/ovn_kubernetes.go:684]: I would change 
```
// It is note that we must disable IPsec temporarily in OVN so that IPsec can be upgraded into latest IPsec daemonsets.
// When IPsec is configured to be disabled, then ensure ovnkube-node pods are rolled out first which configures OVN/OVS to disable IPsec.
// This would give enough room for OVS IPsec monitor script to disble IPsec on the dataplane. After this it is safe to remove IPsec daemonset
// pods.
```
to just
```
The following considerations need to be made for all the procedures:
- IPsec must be disabled in OVN before the IPsec daemon sets are disabled.
- IPsec legacy daemon sets need to be disabled before enabling the updated IPsec daemons sets
```

File: pkg/network/ovn_kubernetes.go
Comment (jcaamano) [inline pkg/network/ovn_kubernetes.go:2542]: So looking at this, I think we should set the annotation in the pod spec within the daemon set as well.
Then you only need a single method instead of three that receives the pods spec `isOVNIPsecNotActive(pod *corev1.Pod)`. You then pass in to this method the pod spec from the daemon set or the stateful set depending on the version you are upgrading from.

File: pkg/network/ovn_kubernetes.go
Comment (jcaamano) [inline pkg/network/ovn_kubernetes.go:2507]: This should be `annotations["release.openshift.io/version"] < 4.15`

Use `isVersionGreaterThanOrEqualTo` for the comparison

File: pkg/network/ovn_kubernetes.go
Comment (jcaamano) [inline pkg/network/ovn_kubernetes.go:676]: replace `ipsecConfig: {} is set` with `ipsecConfig is set` throughout this text block

File: pkg/network/ovn_kubernetes.go
Comment (jcaamano) [inline pkg/network/ovn_kubernetes.go:677]: typo: `render`

File: pkg/network/ovn_kubernetes.go
Comment (jcaamano) [inline pkg/network/ovn_kubernetes.go:676]: replace `networks operator`with `network operator` throughout this text block

File: pkg/network/ovn_kubernetes.go
Comment (jcaamano) [inline pkg/network/ovn_kubernetes.go:680]: replace `.` with `,`

File: pkg/network/ovn_kubernetes.go
Comment (jcaamano) [inline pkg/network/ovn_kubernetes.go:701]: `change to them`

File: pkg/network/ovn_kubernetes.go
Comment (jcaamano) [inline pkg/network/ovn_kubernetes.go:722]: `rolled`

File: pkg/network/ovn_kubernetes.go
Comment (jcaamano) [inline pkg/network/ovn_kubernetes.go:1547]: `what is`

File: pkg/network/ovn_kubernetes.go
Comment (jcaamano) [inline pkg/network/ovn_kubernetes.go:1584]: `what is`

File: pkg/network/ovn_kubernetes.go
Comment (jcaamano) [inline pkg/network/ovn_kubernetes.go:1731]: `exist, so this is an upgrade from 4.14.

File: pkg/network/ovn_kubernetes.go
Comment (jcaamano) [inline pkg/network/ovn_kubernetes.go:2042]: `are already`

File: pkg/network/ovn_kubernetes.go
Comment (jcaamano) [inline pkg/network/ovn_kubernetes.go:2062]:
```
// isIPsecMachineConfigActive returns true if both master and worker's machine config pool are ready with
// ipsec machine config extension rolled out, otherwise returns false.
```

File: pkg/platform/platform.go
Comment (jcaamano) [inline pkg/platform/platform.go:136]: `are not supported there`

File: pkg/platform/platform.go
Comment (jcaamano) [inline pkg/platform/platform.go:144]: `so find it by using the label and looking for the extension`

File: pkg/network/ovn_kubernetes.go
Comment (jcaamano) [inline pkg/network/ovn_kubernetes.go:2053]: Probably compare the api version to `operv1.GroupVersion.String()` as well

**Review by jcaamano — COMMENTED** [review]
/appove
/lgtm
/hold

I just have small spelling nits plus an ask for `containsNetworkOwnerRef`

@kyrtapz could you drive this home? just help @pperiyasamy ensuring CI is ok?

**Comment by jcaamano** [comment]
@pperiyasamy 

It should be desirable to have some level of e2e CI coverage for upgrades as well.
Also I think we need QA to test upgrades from 4.13 as we have a lot going on there when changing from non-IC to IC.

**Comment by pperiyasamy** [comment]
> It should be desirable to have some level of e2e CI coverage for upgrades as well. Also I think we need QA to test upgrades from 4.13 as we have a lot going on there when changing from non-IC to IC.

@yuvalk is working on ipsec upgrade ci lane which is currently in progress [here](https://github.com/openshift/release/pull/46432). 
Yes, @huiran0826 would test the PR covering all the scenarios from QA perspective.

File: pkg/network/ovn_kubernetes.go
Comment (yuvalk) [inline pkg/network/ovn_kubernetes.go:709]: we cant disable E-W IPsec during upgrade
it means we are creating data exposure problem/risk for the customer

**Comment by yuvalk** [comment]
please remove 78701798e3478ed1993ad6aaed2d334919d37122
I think it's better ot have it as part of PR #2147

File: bindata/network/ovn-kubernetes/common/80-ipsec-worker-extensions.yaml
Comment (pperiyasamy) [reply bindata/network/ovn-kubernetes/common/80-ipsec-worker-extensions.yaml:7]: IMO it would be good to go with two different files because these rendered as two different MC objects. This is more readable than creating two objects via programmatically.

**Comment by pperiyasamy** [comment]
> please remove [7870179](https://github.com/openshift/cluster-network-operator/commit/78701798e3478ed1993ad6aaed2d334919d37122) I think it's better ot have it as part of PR #2147

We need this commit to fix mentioned [issue](https://github.com/openshift/cluster-network-operator/pull/2087#discussion_r1417479501), may remove this commit once #2147 is merged.

**Comment by yuvalk** [comment]
LGTM
all my comments are non blocking :-)

**Review by trozet — APPROVED** [review]
/lgtm

**Pull request merged**
